Spectra Training

cjmcgill opened the pull request:
This PR adds training for spectra data type. Training uses the SID loss function and metric. The epistemic uncertainty calculation is carried out by a roundrobin SID function. Spectrum region exclusions are enabled through phase feature and phase mask files.

cjmcgill commented:
Debugged the testing problems. Also added wasserstein loss (earthmovers distance) as an alternative loss function and metric.

cjmcgill commented on README.md:
```diff
@@ -246,11 +249,19 @@ The bond-level features are concatenated with the bond feature vectors before th
 
+### Spectra
+
+One of the data types that can be trained with Chemprop is "spectra". Spectra training is different than other datatypes because it considers the predictions of all targets together. Targets for spectra should be provided as the values for the spectrum at a specific position in the spectrum. The loss function for spectra is SID, spectral information divergence. Alternatively, Wasserstein distance ...
```
Good catch. I started just having this as a spectra-specific argument where you could use an alternative spectra loss function but then I changed my mind later, thinking that it might be nice to have the flexibility to institute other loss functions for other dataset types as well. Though not currently implemented for anything else. Will send out updated test scripts.

cjmcgill commented on chemprop/data/utils.py:
```diff
@@ -231,13 +233,27 @@ def get_data(path: str,
-        for feat_path in features_path:
-            features_data.append(load_features(feat_path))  # each is num_data x num_features
+
+        if features_path is not None:
+            for feat_path in features_path:
+                features_data.append(load_features(feat_path))  # each is num_data x num_features
+
+        if phase_features_path is not None:
+            phase_features = load_features(phase_features_path)
+            for d_phase in phase_features:
+                if not (d_phase.sum() == 1 and np.count_nonzero(d_phase) == 1):
+                    raise ValueError('Phase features must be one-hot encoded.')
+            features_data.append(phase_features)
+        else:
+            phase_features = None
+
         features_data = np.concatenate(features_data, axis=1)
+
     else:
         features_data = None
+        phase_features = None
```
The intent is for phase features to be loaded in duplicatively as phase features and as an entry in molecule features. I want for this data to get used in two ways: as the key for spectra range exclusion with the mask (needs to stay separate and unscaled) and as molecular features that the model can use to differentiate outputs in different phases.

hesther replied:
I see. If you don't use other molecular features, however, `args.use_input_features` is currently False, so the phase features will not get added to your MPNN output (can easily verify via looking at the dimensions of the model) - so you would need to set `args.use_input_features` to True if you have either phase features or molecular features or both.

cjmcgill replied:
> Later in the code, you refer to the variable phase_features as if it held all features
Where is this?

cjmcgill replied:
Also, there will only be one entry for phase_features_path, it's not a list

hesther replied:
Just tried to reconstruct my thought here - with the new information that the phase features should appear in both molecular features and on their own I think I misread the block: Apart from setting `args.use_input_features`, you should be all good here, sorry for the confusion!

cjmcgill replied:
Okay, I'm changing this if/else tree to be disentangled in a way that should be a lot more clear.

hesther replied:
Yeah, I guess this would be helpful!

cjmcgill commented on chemprop/train/cross_validate.py:
```diff
@@ -151,13 +151,20 @@ def cross_validate(args: TrainArgs,
+                task_scores = scores
                 mean, std = np.nanmean(task_scores), np.nanstd(task_scores)
                 row += [mean, std] + task_scores.tolist()
-            writer.writerow(row)
```
Good catch. I will fix this.

cjmcgill commented:
@hesther I've made changes to address all of your comments except adding in new tests, which I will add after I take a little time to plan them out.

hesther commented:
Just tested - works well now (test_scores.csv is written correctly, and the phase features are now also used as molecular features in all cases). I also like how you restructured the features loading step, it is much easier to understand now!

cjmcgill commented:
@hesther I've added testing for spectra functions.

cjmcgill commented:
There's some kind of weird interaction between the reaction testing and the spectra testing. If I comment out the reaction testing it's fine, but if I leave it uncommented there there's an error for the atom vector size.

hesther commented:
The test of commit 7b2577c failed because you removed `--empty_cache` from the reaction testing (which needs to be there otherwise the graphs of some molecules are not recalculated, which leads to an error in dimensions). If spectra have special features, too, you also need `--empty_cache` for the spectra testing (which would be in the last line of `parameterised.expand)  - am testing this currently on my laptop, will let you know how it goes!

hesther commented:
Ok I See the problem now: The reaction testing block sets the feature dimensions differently than regular chemprop. Since these are globals, they will stay at the changed values. This would also happen for custom atom or bond features (for which we currently don't have testing).

Possible solutions:
- Have the reaction test block always at the very end
- Provide a function to reset the dimensions, and call it if a special argument is provided
- Make chemprop always set the global variables

hesther commented:
Thinking about it, if you trained multiple models in a script, it would probably be cleanest to always reset the globals for each training run automatically. The globals are initialized when the package is loaded (`CACHE_GRAPH`, `SMILES_TO_GRAPH`, `EXTRA_ATOM_FDIM`, `EXTRA_BOND_FDIM`, `REACTION`, `EXPLICIT_H`, `REACTION_MODE`), so if they change in the course of model training (e.g. in reaction mode or when using custom features), they never get reset automatically currently. I am not sure whether just providing an argument would be good enough, since that would put it on the user to correctly reinitialize. 

This could be another PR, though. If you just want to finish this PR, but the `--empty_cache` back in the reaction test block, and rename the test block to come last (e.g. from `test_train_single_task_regression_reaction` to `test_z_train_single_task_regression_reaction`), then the tests will pass.

cjmcgill commented:
@hesther I moved `--empty_cache` so that it is added to all test args. Once I moved the reaction test to the end with a z, it looks like it passes training now. I'll merge in the preload model PR and then push again. 

As for the global variables, I wonder if we could do this with a class function. Have the globals be attributes of a class and give that class a function for resetting that we can then call. I think I can implement something like this (but in question whether the way I'd do it would be best practice).

cjmcgill commented:
Squashed and rebased onto master. Amazingly, the refactor of `make_predictions` seems not to have changed anything and all the additional code lines are in the same spots.

cjmcgill commented:
Okay the IR spectra were predicted exactly as well here as they were in my old model code, so it passes that ultimate test! @hesther it's ready if you think it's ready to pass review.

hesther reviewed (commented):
Looks very good to me! I also tested preloading the spectra models and discovered something odd:  When making a prediction through the command line, there is a column `epi_unc` in the csv file of the prediction, but when calling the predict function through python, this column is missing from the saved predictions. This only happens for the spectra. Looks to me like an incompatibility with the preloading, maybe?

hesther reviewed (approved):
The `epi_unc` issue was caused by myself, sorry! So, green light to merge this anytime!
